proxy/grpcproxy/cluster.go

```diff
@@ -15,20 +15,99 @@
 package grpcproxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 
 	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	gnaming "google.golang.org/grpc/naming"
 )
 
 type clusterProxy struct {
 	client *clientv3.Client
+	wa     gnaming.Watcher
+	donec  <-chan struct{}
+	umu    sync.RWMutex
+	umap   map[string]struct{}
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// The returned channel is closed when there is grpc-proxy endpoint registered
+// and the client's context is canceled so the 'register' loop returns.
+func NewClusterProxy(c *clientv3.Client, prefix string, addr string, ttl int) (pb.ClusterServer, <-chan struct{}, error) {
+	cp := &clusterProxy{
 		client: c,
 	}
+
+	var watcher gnaming.Watcher
+	var donec <-chan struct{}
+	var err error
+	if prefix != "" {
```

Comment:
All of this probably should be handled in a separate goroutine / the one that is doing the monitoring, instead of waiting to confirm that it could establish a resolver so it can retry on failure.

The proxy can start by returning its own listen address for the member list (which is probably what it should do if there is no registration prefix), then populate it when it gets results from the watcher.